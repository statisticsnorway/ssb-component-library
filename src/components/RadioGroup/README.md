RadioGroup
========

> A wrapper for multiple radio buttons

### Usage

#### HTML

```html
<div class="ssb-radio-group">
    <div class="ssb-title no-margin"><h5>Title</h5></div>
    <div class="ssb-radio">
        <input type="radio" id="option1" value="option1" />
        <label class="radio-label" for="option1">Insert label</label>
    </div>
    <div class="ssb-radio">
        <input type="radio" id="option2" value="option2" />
        <label class="radio-label" for="option2">Insert label</label>
    </div>
</div>
```

#### React

```jsx harmony
<RadioGroup
  header="Header"
  items={
    {
      label: 'Item 1',
      value: 'item1',
    },
    {
      label: 'Item 2',
      value: 'item2',
    },
  }
  onChange={() => callback}
  selectedValue="item2"
/>
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| header | string | Renders a h5 title |
| items | arrayOf(label, value) | Required items for rendering radio buttons |
| onChange | func | Callback function when a value is changed |
| selectedValue | string | Pre selected value |
