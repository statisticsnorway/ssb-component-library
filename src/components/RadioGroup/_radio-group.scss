.ssb-radio-group {
  @include roboto;
  display: inline-block;

  .radio-group-header {
    padding-left: 10px;
  }
}
