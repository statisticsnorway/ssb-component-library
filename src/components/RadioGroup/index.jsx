import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import Title from '../Title';
import RadioButton from '../RadioButton';

const RadioGroup = ({ header, items, onChange, orientation, selectedValue }) => {
	const [selected, updateSelected] = useState(selectedValue);

	useEffect(() => {
		onChange(selected);
	}, [selected]);

	return (
		<div className="ssb-radio-group">
			{header && <Title className="radio-group-header no-margin" size={5}>{header}</Title>}
			<div className={`boxes flex-${orientation}`}>
				{items.map((it, index) => (
					<RadioButton
						key={it.value}
						index={index}
						selected={it.value === selected}
						value={it.value}
						callback={updateSelected}
						disabled={it.disabled}
					>{it.label}
					</RadioButton>
				))}
			</div>
		</div>
	);
};

RadioGroup.defaultProps = {
	onChange: () => {},
	orientation: 'column',
};

RadioGroup.propTypes = {
	header: PropTypes.string,
	items: PropTypes.arrayOf(PropTypes.shape({
		label: PropTypes.string,
		value: PropTypes.string,
	})).isRequired,
	onChange: PropTypes.func,
	orientation: PropTypes.oneOf([
		'column',
		'row',
	]),
	selectedValue: PropTypes.string,
};

export default RadioGroup;
