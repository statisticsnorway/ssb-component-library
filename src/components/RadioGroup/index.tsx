import React, { useEffect, useState } from 'react'
import RadioButton from '../RadioButton'
import classNames from '../../utils/utils'

interface Item {
  id?: string
  label: string
  value: string | number
  disabled?: boolean
}

interface RadioGroupProps {
  className?: string
  groupName?: string
  header?: string
  items: Item[]
  onChange?: (value: string) => void
  orientation?: 'column' | 'row'
  selectedValue?: string | number
}

const RadioGroup: React.FC<RadioGroupProps> = ({
  className,
  groupName,
  header,
  items,
  onChange = () => {},
  orientation = 'column',
  selectedValue,
}) => {
  const [selected, setSelected] = useState<string | number>(selectedValue || '')

  useEffect(() => {
    onChange(selected.toString())
  }, [selected])

  return (
    <div className={classNames('ssb-radio-group', className)}>
      {header && <div className='radio-group-header'>{header}</div>}
      <div className={`boxes flex-${orientation}`}>
        {items.map((it: Item) => (
          <RadioButton
            id={it.id}
            key={it.value}
            selected={it.value === selected}
            value={it.value}
            name={groupName || header}
            callback={setSelected}
            disabled={it.disabled}
          >
            {it.label}
          </RadioButton>
        ))}
      </div>
    </div>
  )
}

export default RadioGroup
