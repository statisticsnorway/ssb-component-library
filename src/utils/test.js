import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
