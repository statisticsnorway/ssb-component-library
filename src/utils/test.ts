import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
