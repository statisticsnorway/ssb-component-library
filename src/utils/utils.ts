// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
