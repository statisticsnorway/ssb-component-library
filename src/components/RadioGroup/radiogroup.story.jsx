import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import RadioGroup from './index';

const radioItems = [
	{
		label: 'Item 1',
		value: 'item1',
	}, {
		label: 'Item 2',
		value: 'item2',
	}, {
		label: 'Item 3',
		value: 'item3',
		disabled: 'true',
	},
];

storiesOf('RadioGroup', module).addDecorator(centered)
	.add('Column', () => (
		<RadioGroup header="Header" items={radioItems} />
	)).add('Row', () => (
		<RadioGroup header="Header" items={radioItems} orientation="row" />
	));
