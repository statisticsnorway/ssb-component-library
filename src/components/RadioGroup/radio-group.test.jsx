import React from 'react'
import { render } from '../../utils/test'
import RadioGroup from './index'

const radioItems = [
  {
    label: 'Item 1',
    value: 'item1',
  },
  {
    label: 'Item 2',
    value: 'item2',
  },
  {
    label: 'Item 3',
    value: 'item3',
    disabled: true,
  },
]

describe('RadioGroup component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(<RadioGroup header='Header' items={radioItems} />)
    expect(asFragment()).toMatchSnapshot()
  })
  test('Toggles classNames correctly', () => {
    const { asFragment } = render(<RadioGroup header='Header' items={radioItems} orientation='row' />)
    expect(asFragment()).toMatchSnapshot()
  })
})
