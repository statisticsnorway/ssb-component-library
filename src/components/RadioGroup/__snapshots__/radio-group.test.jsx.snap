// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-radio-group"
  >
    <div
      class="radio-group-header"
    >
      Header
    </div>
    <div
      class="boxes flex-column"
    >
      <div
        class="ssb-radio"
      >
        <input
          id="item1"
          name="Header"
          tabindex="0"
          type="radio"
          value="item1"
        />
        <label
          class="radio-label"
          for="item1"
        >
          Item 1
        </label>
      </div>
      <div
        class="ssb-radio"
      >
        <input
          id="item2"
          name="Header"
          tabindex="0"
          type="radio"
          value="item2"
        />
        <label
          class="radio-label"
          for="item2"
        >
          Item 2
        </label>
      </div>
      <div
        class="ssb-radio"
      >
        <input
          disabled=""
          id="item3"
          name="Header"
          tabindex="0"
          type="radio"
          value="item3"
        />
        <label
          class="radio-label"
          for="item3"
        >
          Item 3
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RadioGroup component Toggles classNames correctly 1`] = `
<DocumentFragment>
  <div
    class="ssb-radio-group"
  >
    <div
      class="radio-group-header"
    >
      Header
    </div>
    <div
      class="boxes flex-row"
    >
      <div
        class="ssb-radio"
      >
        <input
          id="item1"
          name="Header"
          tabindex="0"
          type="radio"
          value="item1"
        />
        <label
          class="radio-label"
          for="item1"
        >
          Item 1
        </label>
      </div>
      <div
        class="ssb-radio"
      >
        <input
          id="item2"
          name="Header"
          tabindex="0"
          type="radio"
          value="item2"
        />
        <label
          class="radio-label"
          for="item2"
        >
          Item 2
        </label>
      </div>
      <div
        class="ssb-radio"
      >
        <input
          disabled=""
          id="item3"
          name="Header"
          tabindex="0"
          type="radio"
          value="item3"
        />
        <label
          class="radio-label"
          for="item3"
        >
          Item 3
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
